vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -64,25 +65,71 @@ import dateFnsIsValid from 'date-fns/isValid';
             throw new Error('Array of custom date formats is null or empty');
           }
 
+          function _shortenFourDigitYearFormat(format) {
+            if (format.includes('yyyy') && !format.includes('yyyyy')) {
+              return format.replace('yyyy', 'yy');
+            }
+            if (format.includes('YYYY') && !format.includes('YYYYY')) {
+              return format.replace('YYYY', 'YY');
+            }
+            return undefined;
+          }
+
+          function _isShortFormat(format) {
+            if (format.includes('y')) {
+              return !format.includes('yyy');
+            }
+            if (format.includes('Y')) {
+              return !format.includes('YYY');
+            }
+            return false;
+          }
+
           function formatDate(dateParts) {
             const format = formats[0];
-            const date = datepicker._parseDate(`${dateParts.year}-${dateParts.month + 1}-${dateParts.day}`);
+            const date = _parseDate(`${dateParts.year}-${dateParts.month + 1}-${dateParts.day}`);
 
             return dateFnsFormat(date, format);
           }
 
           function parseDate(dateString) {
+            const referenceDate = _getReferenceDate();
             for (let format of formats) {
-              const date = dateFnsParse(dateString, format, new Date());
+              // We first try to match the date with the shorter version.
+              const shorterFormat = _shortenFourDigitYearFormat(format);
+              if (shorterFormat) {
+                const shorterFormatDate = dateFnsParse(dateString, shorterFormat, referenceDate);
+                if (dateFnsIsValid(shorterFormatDate)) {
+                  let yearValue = shorterFormatDate.getFullYear();
+                  // The last parsed year check handles the case when a date with an actual year value is provided
+                  // with zero padding, but then got reformatted without the zeroes and parsed again.
+                  if (this._lastParsedYear && yearValue == this._lastParsedYear % 100) {
```

Comment:
```suggestion
                  if (datePicker.$connector._lastParsedYear && yearValue === datePicker.$connector._lastParsedYear % 100) {
```
Using `this` here is kind of unpredictable, as it is not obvious what it is going to refer to when the function is called.

For comparing numbers prefer to use strict equality.